tools/instant-bridge/bidder/bidder.go

```diff
@@ -0,0 +1,316 @@
+package bidder
+
+import (
+	"context"
+	"errors"
+	"io"
+	"log/slog"
+	"math/big"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	bidderapiv1 "github.com/primev/mev-commit/p2p/gen/go/bidderapi/v1"
+	debugapiv1 "github.com/primev/mev-commit/p2p/gen/go/debugapi/v1"
+	notificationsapiv1 "github.com/primev/mev-commit/p2p/gen/go/notificationsapi/v1"
+)
+
+const (
+	epochNotificationTopic = "epoch_validators_opted_in"
+	slotDuration           = 12 * time.Second
+)
+
+var (
+	ErrNoEpochInfo       = errors.New("no epoch info available")
+	ErrNoSlotInNextEpoch = errors.New("no slot available in next epoch")
+)
+
+var nowFunc = time.Now
+
+type slotInfo struct {
+	slot      uint64
+	startTime time.Time
+	blsKey    string
+}
+
+type epochInfo struct {
+	epoch     uint64
+	startTime time.Time
+	slots     []slotInfo
+}
+
+type BlockNumberGetter interface {
+	BlockNumber(ctx context.Context) (uint64, error)
+}
+
+type BidderClient struct {
+	logger              *slog.Logger
+	bigWg               sync.WaitGroup
+	bidderClient        bidderapiv1.BidderClient
+	topologyClient      debugapiv1.DebugServiceClient
+	notificationsClient notificationsapiv1.NotificationsClient
+	nextEpoch           atomic.Pointer[epochInfo]
+	blkNumberGetter     BlockNumberGetter
+}
+
+func NewBidderClient(
+	logger *slog.Logger,
+	bidderClient bidderapiv1.BidderClient,
+	topologyClient debugapiv1.DebugServiceClient,
+	notificationsClient notificationsapiv1.NotificationsClient,
+	blkNumberGetter BlockNumberGetter,
+) *BidderClient {
+	return &BidderClient{
+		logger:              logger,
+		bidderClient:        bidderClient,
+		topologyClient:      topologyClient,
+		notificationsClient: notificationsClient,
+		blkNumberGetter:     blkNumberGetter,
+	}
+}
+
+func (b *BidderClient) Start(ctx context.Context) <-chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		lastMsg := nowFunc()
+	RESTART:
+		sub, err := b.notificationsClient.Subscribe(ctx, &notificationsapiv1.SubscribeRequest{
+			Topics: []string{epochNotificationTopic},
+		})
+		if err != nil {
+			b.logger.Error("failed to subscribe to notifications", "error", err)
+			return
+		}
+
+		if time.Since(lastMsg) > 15*time.Minute {
+			b.logger.Error("no messages received for 15 minutes, closing subscription")
+			return
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				b.logger.Info("context done")
+				return
+			default:
+			}
+
+			msg, err := sub.Recv()
+			if err != nil {
+				b.logger.Error("failed to receive message", "error", err)
+				goto RESTART
+			}
+
+			lastMsg = nowFunc()
+
+			b.logger.Debug("received message", "msg", msg)
+
+			if msg.Topic != epochNotificationTopic {
+				b.logger.Error("unexpected topic", "topic", msg.Topic)
+				continue
+			}
+
+			epoch, err := parseEpochInfo(msg)
+			if err != nil {
+				b.logger.Error("failed to parse epoch info", "error", err, "msg", msg)
+				continue
+			}
+
+			b.nextEpoch.Store(epoch)
```

Comment:
As mentioned on the call we likely need to check the proposer slot assignments _during the epoch in question_, as proposer selection is only stable within the context of the current epoch. See specification [here](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/validator.md#validator-assignments) and [here](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/validator.md#lookahead)

Reply:
Updated the notification on mev-commit to only send current epoch information. Also reworded this to current epoch. So now we should only get current epoch related information.